For dealii 9, use chunk manifold only, otherwise set correct boundary objects

anne-glerum opened the pull request:
See discussion in #1361. @gassmoeller Could you try it with dealii version 9?

bangerth commented:
/run-tests

bangerth commented on include/aspect/compat.h:
```diff
@@ -65,6 +65,347 @@ namespace aspect
   };
 }
+
+#include <deal.II/grid/tria_boundary_lib.h>
```
Please put header files at the top of the file.

gassmoeller replied:
I thought for compat.h we always included headers at the position of the code that uses them to remember to remove them. When they are at the top they might stay around. Is that not correct?

bangerth replied:
Well, I guess we can do that.

bangerth commented on include/aspect/compat.h:
```diff
@@ -65,6 +65,347 @@ namespace aspect
+namespace dealii
+{
+  template <int dim>
+  class SpecialConeBoundary : public StraightBoundary<dim>
```
how does this class differ from deal.II's class?

bangerth commented on include/aspect/compat.h:
```diff
@@ -65,6 +65,347 @@ namespace aspect
+#include <deal.II/grid/tria_boundary_lib.h>
+
+
+namespace dealii
```
bad idea -- this namespace belongs to deal.II :-) Only put things into the `aspect` namespace.

bangerth commented on include/aspect/compat.h:
```diff
@@ -65,6 +65,347 @@ namespace aspect
+
+}
+namespace dealii
+{
```
same here

bangerth commented on include/aspect/geometry_model/chunk.h:
```diff
@@ -19,11 +19,12 @@
-#ifndef _aspect_geometry_model_chunk_h
-#define _aspect_geometry_model_chunk_h
+#ifndef __aspect__geometry_model_chunk_h
+#define __aspect__geometry_model_chunk_h
```
we really should only use one underscore

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -150,6 +150,28 @@ namespace aspect
 
+     /**
+      * A signal that is called before the computation of tangential boundary
+     * conditions for which normal vectors are needed.
```
indentation of the asterisks is broken

bangerth commented on source/geometry_model/chunk.cc:
```diff
@@ -1,5 +1,5 @@
 /*
-  Copyright (C) 2011 - 2016 by the authors of the ASPECT code.
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
```
2017?

bangerth commented on source/geometry_model/chunk.cc:
```diff
@@ -118,6 +166,17 @@ namespace aspect
 
+#if !DEAL_II_VERSION_GTE(9,0,0)
+    template <int dim>
+    void
```
you're going to get linker errors for older deal.II versions because you declare the function but then don't define it. please move the `#if` to only comment out the *body* of the function.

bangerth commented on source/geometry_model/chunk.cc:
```diff
@@ -139,18 +239,101 @@ namespace aspect
+    }
+
+#if !DEAL_II_VERSION_GTE(9,0,0)
+    template <int dim>
```
same issue here

bangerth replied:
unless you have the same `#if` statement in the .h file commenting out the *declaration* of the function.

bangerth commented:
There are some compilation and indentation errors. Can you please take care of them as well?

anne-glerum commented:
I have not updated the test output, because it has changed significantly for those 14 tests that fail. Those failing tests all use chunk geometries, and they are the only tests with chunk geometries.

gassmoeller commented:
This PR also runs with deal.II 9.0, and I get similar changes in the solutions. Could you check for one test or so that the mesh looks reasonable and the solution is reasonable as well. If so we might as well update the tests. We already saw in #1435 how much the change in geometry can affect the solution.

anne-glerum commented:
For comparison, below a figure of the `adiabatic_initial_conditions_chunk.prm` test, left with the new chunk geometry, right the old. The contour indicates a higher temperature blob. On the right, velocities are one magnitude higher and focused on mesh refinement level interfaces instead of arising from the blob area as seen on the left. So (at least in 2d) the solution is reasonable and the mesh as well. 

![screen shot 2017-05-10 at 20 11 42](https://cloud.githubusercontent.com/assets/7631624/25926461/5d8f1c02-35bd-11e7-8455-9b02bc4304fa.png)

gassmoeller commented:
Wow, the new solution looks way better than the old one. Let me know when you updated the tests and want me to take another look.

gassmoeller commented:
Oh and that figure might also look great in the google doc report :smirk:

gassmoeller commented:
sorry for spamming, also add an entry to doc/modules/changes

bangerth commented:
Wow, better by a factor of ten! Very nice!

bangerth commented:
@luca-heltai -- you may be interested in seeing this as well :-)

anne-glerum commented:
Alright, I've updated the tests and added an entry to changes. I'd like it if you have another look at the implementation, as in 3d there is still a problem with velocity along the north and south boundary, see #1361 and the figure below that shows 3 chunks with different radii, but otherwise the same setup (no driving forces present). Velocities along the north boundary are not reduced with stricter solver tolerance, they change with min/max latitude of the chunk but not according to a clear pattern, decrease with increased longitudinal resolution and with inner/outer radius.  
![screen shot 2017-05-07 at 23 42 55](https://cloud.githubusercontent.com/assets/7631624/25930947/047d12b2-35d7-11e7-9b73-8e9246f20e18.png)

bangerth commented:
I'll put it on my list for later today.

In the meantime, there is a merge conflict. Can you see if you can rebase to the current state?

gassmoeller commented:
Very nice! A few style comments, and the merge conflict, but after that this is ready to go. We can close a whole bunch of issues and other pull requests when this is done! Good work :+1:

gassmoeller commented:
Oh weird. The comments that do not seem to make sense in my review are answers to @bangerth's comments above. It seems I can not answer to other comments as part of a new review.

bangerth commented:
It's on my list of things to do to look at this again.

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -151,6 +151,28 @@ namespace aspect
     /**
+     * A signal that is called before the computation of tangential boundary
+    * conditions for which normal vectors are needed.
+    *
```
watch the indentation

anne-glerum replied:
Shouldn't this be picked up by astyle? Neither running doc/indent nor running astyle directly on the file fixes the indentation.

bangerth replied:
Yes, astyle doesn't indent comments :-( You have to do it by hand, unfortunately.

bangerth commented on tests/adiabatic_initial_conditions_chunk/screen-output:
```diff
@@ -5,31 +5,31 @@ Number of degrees of freedom: 1,444 (882+121+441)
    Solving temperature system... 0 iterations.
    Rebuilding Stokes preconditioner...
-   Solving Stokes system... 26+0 iterations.
+   Solving Stokes system... 25+0 iterations.
```
why does this file change? the test uses a spherical shell boundary, but I thought this patch should only apply to the chunk geometry, right?

anne-glerum replied:
The test uses a chunk geometry and tangential velocity boundary conditions on all boundaries but the outer boundary. It is the same test as for which I posted the picture. So I do think the results should change. Or did you mean another test?

bangerth replied:
Oh, I had looked at `adiabatic_initial_conditions` by accident. Thanks for reminding me.

It is a bit of a shame that this patch increases the number of iterations (in further time steps) by so much :-( But it is strictly more correct, so I think we should still go with it.

bangerth commented:
You'll have to rebase because of conflicts, and probably want to merge it all into one commit as well.

anne-glerum commented on include/aspect/compat.h:
```diff
@@ -65,6 +65,73 @@ namespace aspect
+      virtual
+      Tensor<1,dim>
+      normal_vector (const typename Triangulation<dim>::face_iterator &face,
+                     const Point<dim> &p) const;
```
It did not. That would be great if you could do that.

bangerth replied:
Will work on it.

But on a tangent, we only need this class with old deal.II versions anyway, right? In that case, won't the issue go away sometime soon anyway?

anne-glerum replied:
yes

bangerth replied:
Ah, ok. In that case maybe I won't worry about it in deal.II right away. I have a patch running in deal.II, just no test yet. I might take care of it next week getting it upstream then.

anne-glerum commented on doc/modules/changes/20170510_anne-glerum:
```diff
@@ -0,0 +1,6 @@
+Changed: The chunk geometry now removes the manifold id only from the boundaries and sets the correct
+boundary objects for dealii versions prior to version 9. For dealii 9, only the manifold is used,
```
It is not in 8.5 yet, so we include compiler statements to test against deal.II 9 or newer (#if !DEAL_II_VERSION_GTE(9,0,0)).

bangerth replied:
OK, I just wanted to check whether 9.0 really was the correct version number. Thanks for confirming.

gassmoeller commented on include/aspect/geometry_model/chunk.h:
```diff
@@ -259,6 +272,10 @@ namespace aspect
             push_forward(const Point<dim> &chart_point) const;
 
             virtual
+            DerivativeForm<1, dim, dim>
```
Documentation please.

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -99,7 +147,7 @@ namespace aspect
             // See 2D case
             if (output_vertex[1] < 0.0)
-              if (output_vertex[1] < point1_lon - std::abs(point1_lon)*std::numeric_limits<double>::epsilon())
+              if (output_vertex[1] < point1_lon - 100 * std::abs(point1_lon)*std::numeric_limits<double>::epsilon())
```
Why 100? is that necessary? If so describe in a comment why, if not make it the same as in the other function.

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -131,6 +190,47 @@ namespace aspect
                                                  point2,
                                                  true);
 
+      // At this point, all boundary faces have their correct boundary
```
I think all of this can go into an `#ifdef` as well, correct?

gassmoeller commented on source/geometry_model/chunk.cc:
```diff
@@ -139,18 +239,101 @@ namespace aspect
+          static const SpecialConeBoundary<dim> boundary_cone_north(0.0,north_radius,center,north);
+          static const SpecialConeBoundary<dim> boundary_cone_south(0.0,south_radius,center,south);
+          //static const ConeBoundary<dim> boundary_cone_north(0.0,north_radius,center,north);
+          //static const ConeBoundary<dim> boundary_cone_south(0.0,south_radius,center,south);
```
these commented lines can go

gassmoeller commented:
Ok, thanks. I am going to merge this PR to get the improved model in. Nevertheless, we should discuss the remaining velocity anomalies with @bangerth.
